include/picongpu/particles/shapes/PCS.hpp

```diff
@@ -86,63 +110,108 @@ namespace picongpu
                          *       |  1/96 * (55 + 4 * x * (5 - 2 * x * (15 + 2 * x * (-5 + x)))) if 1/2 <= |x| < 3/2
                          *       |
-                         *       |  1/384 * (5 - 2 * x)^4                                       if 3/2 <= |x| < 5/2
+                         *       |  1/384 * (5 - 2*x)^4                                         if 3/2 <= |x| < 5/2
+                         *       |
+                         *       |  0                                                           otherwise
                          *       -
                          */
                         float_X const abs_x = math::abs(x);
 
-                        bool const below_2nd_radius = abs_x < 1.5_X;
-                        bool const below_1st_radius = abs_x < 0.5_X;
+                        bool const below_max = abs_x < 2.5_X;
 
-                        float_X const rad1 = ff_1st_radius(abs_x);
-                        float_X const rad2 = ff_2nd_radius(abs_x);
-                        float_X const rad3 = ff_3rd_radius(abs_x);
+                        float_X const onSupport = ChargeAssignmentOnSupport()(abs_x);
 
-                        float_X result = rad3;
-                        if(below_1st_radius)
-                            result = rad1;
-                        else if(below_2nd_radius)
-                            result = rad2;
+                        float_X result(0.0);
+                        if(below_max)
+                            result = onSupport;
 
                         return result;
                     }
+
+                    /** Creates an array with assignment values.
+                     *
+                     * @param pos particle position relative to the assignment cell range [-0.5;1.5)
+                     * @param isOutOfRange True if pos in range [-0.5;1.5)
+                     * @return Array with precomputed assignment values.
+                     */
+                    HDINLINE auto shapeArray(float_X const xx, bool const isOutOfRange) const
+                    {
+                        float_X x = isOutOfRange ? xx - 1.0_X : xx;
+
+                        auto shapeValues = detail::PCS::shapeArray<support + 1>(x);
+
+                        // Update value so that a particle can be out of range without using lmem/local memory on GPUs
+                        // because of dynamic indexing into an array located in registers.
+                        shapeValues[5] = isOutOfRange ? shapeValues[4] : 0.0_X;
+                        shapeValues[4] = isOutOfRange ? shapeValues[3] : shapeValues[4];
+                        shapeValues[3] = isOutOfRange ? shapeValues[2] : shapeValues[3];
+                        shapeValues[2] = isOutOfRange ? shapeValues[1] : shapeValues[2];
+                        shapeValues[1] = isOutOfRange ? shapeValues[0] : shapeValues[1];
+                        shapeValues[0] = isOutOfRange ? 0.0_X : shapeValues[0];
```

Comment:
Could these be expressed as an `if` statement, which imo would be more readable then

Reply:
This is the magic trick of the optimization. It is required that we use the ternary operator to allow the compiler to track all memory access and avoid that lmem will be used for the array.